autopilot/contractor.go

```diff
@@ -546,8 +546,30 @@ func (c *contractor) candidateHosts(cfg Config, wanted int) ([]consensus.PublicK
 		wanted = len(hosts)
 	}
 
-	// TODO: score each host and return selected hosts
-	return nil, nil
+	// score each host
+	scores := make([]float64, len(hosts))
+	for i, h := range hosts {
+		scores[i] = newHostScore(host{h, rhpv2.Contract{}}).
+			withAgeScore().
+			withCollateralScore(cfg).
+			withInteractionScore().
+			withSettingsScore(cfg).
+			withUptimeScore().
+			withVersionScore().
+			finalize()
+	}
+
+	// select hosts
+	var selected []consensus.PublicKey
+	for len(selected) < wanted {
+		i := randSelectByWeight(scores)
```

Comment:
it makes me slightly unease to hide a double-nested `for` loop here...but I think *n* is small enough that it won't be an issue in practice. If it does start showing up in perf profiles, it will be easy enough to inline the loop and eliminate duplicated work.